### PR TITLE
Calculate approximate number of accessed unique pages to estimate working set size at compute

### PR DESCRIPTION
**Pull request by knizhnik**
## Describe your changes

Calculate number of unique page accesses at compute.
It can be used to estimate working set size and adjust cache size (shared_buffers or local file cache).

Approximation is made using HyperLogLog algorithm.
It is performed by local file cache and so is available only when local file cache is enabled.

This calculation doesn't take in account access to the pages present in shared buffers, but includes pages available in local file cache.

This information can be retrieved using `approximate_working_set_size(reset bool)` function from  `neon` extension.
`reset` parameter can be used to reset statistic and so collect unique accesses for the particular interval.

Below is an example of estimating working set size after `pgbench -c 10 -S -T 100 -s 10`:

```
postgres=# select approximate_working_set_size(false);
 approximate_working_set_size 
------------------------------
                        19052
(1 row)

postgres=# select pg_table_size('pgbench_accounts')/8192;
 ?column? 
----------
    16402
(1 row)
```

## Checklist before requesting a review
- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

### CODE REVIEW

**Comment by vadim2404** [comment]
How can it be used, and by whom?

File: pgxn/neon/file_cache.c
Comment (vadim2404) [inline pgxn/neon/file_cache.c:102]: It's not true. `1 << 32` is `pow(2,32)`. Here you have `2 ** 32 - 1`. I assume that it was the idea, but the name misleads

**Comment by knizhnik** [comment]
> How can it be used, and by whom?

My original intention was that it an be used by autoscaler to determine size of local file cache needed for tenant to provide maimal performance.
Certainly we can use eager strategy: specify large local cache size, let tenant fill it and it there is not enough free space, then shrink cache size. But knowledge of working set size can allows smarter scheduling of tenant between nodes.
Originally discussed here: https://neondb.slack.com/archives/C03TN5G758R/p1680158640939199

**Comment by sharnoff** [comment]
I really like the idea of having this available. Actually using it for autoscaling will probably require some careful thought, but it seems like a clean way to cover the last big area (CPU/memory usage, etc.) that we don't have any introspection into.

**Comment by vadim2404** [comment]
I'm not sure I understood the idea of 100% of how this info can be used for autoscaling. 

It shows the data size, but we must scale based on the load profile.

Imagine the situation with the situation.

The user spun up the read instance for OLAP queries. The data size is enormous. The user runs read queries from time to time. How should we understand to downscale the instance if it was upscaled because of data size?

About which thing I thought? It's related to our network latency. If users execute a lot of queries, but the CPU is pretty idle, it seems that we need to upscale the memory. When such stats go down, it's the reason to downscale. However, it's also a kind of heuristic model.

**Comment by knizhnik** [comment]
> I'm not sure I understood the idea of 100% of how this info can be used for autoscaling.
> 
> It shows the data size, but we must scale based on the load profile.
> 
> Imagine the situation with the situation.
> 
> The user spun up the read instance for OLAP queries. The data size is enormous. The user runs read queries from time to time. How should we understand to downscale the instance if it was upscaled because of data size?
> 
> About which thing I thought? It's related to our network latency. If users execute a lot of queries, but the CPU is pretty idle, it seems that we need to upscale the memory. When such stats go down, it's the reason to downscale. However, it's also a kind of heuristic model.

It shows not "data size", but working set size. It can be much smaller than database size.
And definitely it can not be the only criteria for autoscaling. We should also looks at CPU and memory usage.

Sorry if my explanation were not clear.
I tried to do it once again.

Right now for large enough data set (which doesn't fir in compute shared buffers),
the performance bottleneck is compute<->page server round-trip.
If prefetch can not be use then compute most of the time will spent in waiting for response for get_page@lsn
So if autoscaler looks just at CPU/disk and memory usage - it will not see any need to up-scale this node: 
CPU usage is almost 0, disk activity is low, extra memory is not requested. Autoscaler can make an assumption
than there is low activity and even try to downscale this node. While customers complain about bad performance.

There are two ways to avoid such scenario:
1. Large (overcommitted) local file cache. This cache will be filled by loaded pages and sooner or later one of two things will happen:
- We keep all working set in local file cache and performance is comparable with standalone Postgres.
- Size of local storage is not enough to keep such large local file cache and so we have to shrink it.
2. Estimate working set size and based o this size make one of the decisions:
- Extend local file cache to contain all working set
- Move compute to another node with larger enough free local storage space
- Do not try to extend local file cache size because in any case there is no chance to fit in local storage.

Scenario 2 is more flexible than 1. we can do smart planning and make a decision before we are local of local storage free space.
As in your example with  OLAP  query we may decide that there is no sense to try to extend local file cache to fit all huge database. While in case of pgbench we can extend local file cache to fit the whole database (which is not so large - hundreds of Gb)    to achive the best performance.

**Comment by vadim2404** [comment]
This explanation helped me more. Thanks for your answer. 

What to do with RAM (and `shared_buffers`) if we will be completely relying on file local cache?

File: pgxn/neon/file_cache.c
Comment (knizhnik) [reply pgxn/neon/file_cache.c:20]: No

**Comment by knizhnik** [comment]
> This explanation helped me more. Thanks for your answer.
> 
> What to do with RAM (and `shared_buffers`) if we will be completely relying on file local cache?

Well, we can't extend shared buffers without restarting postgres (this was one of the main reasons of trying local file cache instead). And it is not actually needed. If shared buffers has some reasonable size (i.e. 128Mb not default 1Mb), then it is enough to fir catalog relations and together with local file cache we can provide performance comparable with standalone Postgres.

Concerning memory it is more interesting question.  More memory may be needed
1. To handle more connections (`workmem` multiplied y number of active connections)
2. Increase `workmem` and so use more efficient query plans (do sorting in memory, hash joins,...)
3. Maintain temporary resutl sets in memory without spilling to disk
4. Use memory for OS file cache
Please notice that local file cache is most efficient if it fits in  OS cache. So no disk reads are needed at all.
Average SSD access time is about 0.1msec. It is comparable with network latency within one AZ. Memory access speed is several orders of magnitude larger.

So local file cache should be used in conjunction with increasing node memory.

**Review by Bodobolero — APPROVED** [review]
This function is very useful to end users and customers to determine the working set of a specific workload and determine the number of vCPU for the compute node.

Using the hyperloglog is very smart and saves a lot of memory.

**Comment by Bodobolero** [comment]
@knizhnik Pls notify me once this PR is merged - I will add a test case then (since the base has evolved with new versions of neon extension it doesn't make sense to prepare the test case before you have merged the functionality)

**Comment by knizhnik** [comment]
Replaced with https://github.com/neondatabase/neon/pull/6935

**Pull request closed without merging**
